### PR TITLE
Fix resolver race

### PR DESCRIPTION
**Pull request by Random-Liu**
Port https://github.com/containerd/containerd/pull/2276 to the resolver package we are using.

Fixes https://github.com/containerd/cri/issues/722.

### CODE REVIEW

File: pkg/containerd/resolver/resolver.go [hunk 1]
```diff
@@ -250,12 +251,12 @@ func (r *containerdResolver) Pusher(ctx context.Context, ref string) (remotes.Pu
...
 
-	client   *http.Client
-	useBasic bool
-	username string
-	secret   string
+	client           *http.Client
+	useBasic         bool
+	username, secret string
+	token            string
+	mu               sync.Mutex
```
Comment (mikebrow) [inline pkg/containerd/resolver/resolver.go]: // lock used to protect access to token..

File: pkg/containerd/resolver/resolver.go [hunk 2]
```diff
@@ -300,6 +301,23 @@ func (r *containerdResolver) base(refspec reference.Spec) (*dockerBase, error) {
 	}, nil
 }
 
+func (r *dockerBase) getToken() string {
```
Comment (mikebrow) [inline pkg/containerd/resolver/resolver.go]: // getToken returns token value protected by mu mutex

File: pkg/containerd/resolver/resolver.go [hunk 3]
```diff
@@ -300,6 +301,23 @@ func (r *containerdResolver) base(refspec reference.Spec) (*dockerBase, error) {
...
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.token
+}
+
+func (r *dockerBase) setToken(token string) bool {
```
Comment (mikebrow) [inline pkg/containerd/resolver/resolver.go]: ditto

**Comment by Random-Liu** [comment]
@mikebrow I prefer keeping the change identical to containerd/containerd#2276 to make it easy to port future changes. :)

We should consolidate this in the future anyway https://github.com/containerd/cri/issues/723.

**Pull request merged**